lib/wallet/src/Cardano/Wallet/Read/Primitive/Block/Header.hs

```diff
@@ -0,0 +1,152 @@
+{-# LANGUAGE ApplicativeDo #-}
+{-# LANGUAGE DataKinds #-}
+{-# LANGUAGE GADTs #-}
+{-# LANGUAGE LambdaCase #-}
+
+module Cardano.Wallet.Read.Primitive.Block.Header
+    ( getBlockHeader
+    , primitiveBlockHeader
+    )
+where
+
+import Prelude hiding
+    ( (.)
+    )
+
+import Cardano.Crypto.Hash.Class
+    ( hashToBytes
+    )
+import Cardano.Ledger.Crypto
+    ( StandardCrypto
+    )
+import Cardano.Wallet.Read
+    ( Block
+    , ConsensusBlock
+    , fromConsensusBlock
+    )
+import Cardano.Wallet.Read.Block.BlockNo
+    ( BlockNo (..)
+    , getEraBlockNo
+    )
+import Cardano.Wallet.Read.Block.HeaderHash
+    ( HeaderHash (..)
+    , HeaderHashT
+    , PrevHeaderHash (..)
+    , PrevHeaderHashT
+    , getEraHeaderHash
+    , getEraPrevHeaderHash
+    )
+import Cardano.Wallet.Read.Block.SlotNo
+    ( SlotNo (..)
+    , getEraSlotNo
+    )
+import Cardano.Wallet.Read.Eras
+    ( EraFun
+    , applyEraFun
+    , extractEraValue
+    )
+import Cardano.Wallet.Read.Eras.EraFun
+    ( EraFun (..)
+    , EraFunK (..)
+    )
+import Control.Category
+    ( (.)
+    )
+import Data.Coerce
+    ( coerce
+    )
+import Data.Quantity
+    ( Quantity (..)
+    )
+import Generics.SOP
+    ( K (..)
+    )
+import Ouroboros.Consensus.Byron.Ledger.Block
+    ( ByronHash (..)
+    )
+import Ouroboros.Consensus.Shelley.Protocol.Abstract
+    ( ShelleyHash (..)
+    )
+
+import qualified Cardano.Crypto.Hashing as CC
+import qualified Cardano.Protocol.TPraos.BHeader as SL
+import qualified Cardano.Wallet.Primitive.Types as W
+import qualified Cardano.Wallet.Primitive.Types.Hash as W
+import qualified Ouroboros.Network.Block as O
+
+fromByronHash :: ByronHash -> W.Hash tag
+fromByronHash = W.Hash . CC.hashToBytes . unByronHash
+
+-- | Get a wallet primitive block header from a ledger block
+getBlockHeader :: W.Hash "Genesis" -> ConsensusBlock -> W.BlockHeader
+getBlockHeader gp =
+    extractEraValue
+        . applyEraFun (primitiveBlockHeader gp)
+        . fromConsensusBlock
+
+-- | Compute a wallet primitive block header from a ledger
+primitiveBlockHeader :: W.Hash "Genesis" -> EraFun Block (K W.BlockHeader)
+primitiveBlockHeader gp =
+    fromEraFunK $ do
+        slotNo <- fromSlotNo <$> EraFunK getEraSlotNo
+        blockNo <- fromBlockNo <$> EraFunK getEraBlockNo
+        headerHash <- EraFunK (primitiveHash . getEraHeaderHash)
+        -- TODO: what is the Maybe for ? remove it !
```

Comment:
```suggestion
        -- TODO: what is the Maybe for ? remove it !
```

Answer: In the world of `W.Block`, the genesis information is encoded as a `W.Block` — there is a `block0` floating around in other parts of the wallet. For this "block", the `prevHeaderHash` is `Nothing`.

Reply:
Ah ok, found the place where it's nothing. Lost something ? Look in compatibility module !